tsdb/importer/import.go

```diff
@@ -0,0 +1,305 @@
+	"github.com/prometheus/prometheus/pkg/textparse"
+	"github.com/prometheus/prometheus/tsdb"
+	tsdb_labels "github.com/prometheus/prometheus/tsdb/labels"
+)
+
+// Implementing the error interface to create a
+// constant, which cannot be overridden.
+// https://dave.cheney.net/2016/04/07/constant-errors
+type Error string
+
+func (e Error) Error() string {
+	return string(e)
+}
+
+// This error is thrown when we try to merge/add blocks to an existing TSDB instance,
+// and the new blocks have a time overlap with the current blocks.
+const OverlappingBlocksError = Error("blocks overlap with blocks currently in DB")
+
+// Duration of a block in milliseconds
+const BlockDuration = 2 * 60 * 60 * 1000
+
+type timestamp = int64
+
+type metricSample struct {
+	TimestampMs timestamp
+	Value       float64
+	Labels      tsdb_labels.Labels
+}
+
+// ImportFromFile imports data from a file formatted according to the Prometheus exposition format,
+// converts it into block(s), and places the newly created block(s) in the
+// TSDB DB directory, where it is treated like any other block.
+func ImportFromFile(filePath string, contentType string, dbPath string, skipTimestampCheck bool, logger log.Logger) error {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+
+	tmpDbDir, err := ioutil.TempDir("", "importer")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpDbDir)
+
+	dbMint, dbMaxt, err := getDbTimeLimits(dbPath)
+	if err != nil {
+		return err
+	}
+
+	f, err := os.Open(filePath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	bytes, err := ioutil.ReadAll(f)
+	if err != nil {
+		return err
+	}
+
+	blockPaths, err := pushMetrics(bytes, contentType, tmpDbDir, dbMint, dbMaxt, skipTimestampCheck, logger)
+	if err != nil {
+		return err
+	}
+
+	level.Info(logger).Log("msg", "blocks created", "blockPaths", blockPaths)
+
+	err = copyToDatabase(tmpDbDir, dbPath)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// pushMetrics parses metrics formatted in the Prometheus exposition format,
+// and creates corresponding blocks.
+// Returns paths to the newly created blocks, and error.
+func pushMetrics(b []byte, contentType string, dbPath string, dbMint, dbMaxt timestamp, skipTimestampCheck bool, logger log.Logger) ([]string, error) {
+	//var minValidTimestamp timestamp
+	minValidTimestamp := int64(math.MaxInt64)
+	//var maxValidTimestamp timestamp
+	maxValidTimestamp := int64(math.MinInt64)
+
+	var blockPaths []string
+	var currentBucket []*metricSample
+	var startTime time.Time
+	var currentTime time.Time
+	var err error
+	parser := textparse.New(b, contentType)
+	for {
+		var ent textparse.Entry
+		if ent, err = parser.Next(); err != nil {
+			// Error strings are just different enough across packages,
+			// hence this catch-all that just looks for "EOF" in the error
+			// string, and if it finds one, it means that the parsing is complete.
+			if strings.Contains(strings.ToLower(err.Error()), "eof") {
```

Comment:
there should be some proper EOF error which you can match to avoid string matching here.

Reply:
In this, while it does return `io.EOF` on properly reaching the EOF, I've found that the OpenMetricParser throws a different kind of EOF error, when it reaches the end of a word

https://github.com/prometheus/prometheus/blob/1e62435960fcd4991e320fc055de412e0cd7ed9f/pkg/textparse/openmetricsparse.go#L190

Reply:
in that case can match for both or open a PR to make both return the same EOF error

Reply:
I think I can tackle that with a new PR

Reply:
Thanks. Appreciated. 

Please link it to this one and I can review it.

Since this is not an urgent one lets not merge this one until we have a better errors handling

Reply:
That's not an EOF error from the OM parser, that's an unexpected data error meaning the input is invalid.

Reply:
So I've been trying to figure out what's wrong with the test case used here: https://github.com/prometheus/prometheus/pull/5887/files/33cc9bf3428e7221d43189a2949a3751e860ec4f#diff-0518cdc377fbc8a3c23ae09c7be25a4aR75

The OM parser throws an error like so 'Unexpected end of data, EOF' on that test, if switch the check to just look for `io.EOF`. Which is why I wrote this work-around in the first place. I haven't been able to figure out why that happens.

Reply:
That's not valid OpenMetrics input, it's missing the # EOF

Reply:
Thanks, that helped me. I've corrected this to look for the error code `io.EOF`.